### ISSUE TITLE
[RFE][networkd][IPv6] DHCPv6 client should only be triggered a valid RA

### ISSUE THREAD

**Issue by tomty89** [systemd/systemd#3399]
### Submission type
- [ ] Bug report
- [X] Request for enhancement (RFE)
### systemd version the issue has been seen with

> 230
### Used distribution

> Arch Linux

So DHCPv6 client in networkd will be "started" (as in, sd_dhcp6_client_start() is called) in three cases: link_acquire_ipv6_conf(), ndisc_router_handler() and ndisc_handler().

The problem is, it seems to me that DHCPv6 client should not be started in the first (explicitly enabled in .network) and the third case (router discovery timed out).

First, DHCPv6 doesn't seem to work without Router Advertisement anyway. Second, without a valid Router Advertisement (https://tools.ietf.org/html/rfc5175), we have no way to know which mode (i.e. dhcp6_request_address() or not) we should start it in.

You can also see this awkwardness by looking into link_acquire_ipv6_conf() and ndisc_handler(). The former will never call dhcp6_request_address(), and the latter calls it unconditionally.

So I am proposing that we should remove the DHCPv6 starting code in link_acquire_ipv6_conf() and ndisc_handler(). That also means `DHCP=` in .network will be for DHCPv4 only.

I am writing this as an RFE (I mean, instead of a Pull Request) because these are practically harmless (at least in my use case/with my router). Also I am not really some IPv6 expert so I don't know if I've missed something. If I haven't, I do think this should be done to avoid confusion and unnecessary effort for further maintenance.

What do you think? @poettering @keszybz @teg

**Comment by tomty89** [comment]
For the record, I think

```
                assert(link->dhcp6_client);
                assert(in_addr_is_link_local(AF_INET6, (const union in_addr_union*)&link->ipv6ll_address) > 0);
````

and maybe also:

```
                log_link_debug(link, "Acquiring DHCPv6 lease");
````

 should be added to ndisc_router_handler()?

**Comment by tomty89** [comment]
This shows that DHCPv6 does not work without Router Advertisement:

```
[tom@localhost ~]$ cat /etc/systemd/network/enp3s0.network 
[Match]
Name=enp3s0
[Network]
DHCP=yes

[tom@localhost ~]$ grep nameserver /run/systemd/resolve/resolv.conf 
nameserver 192.168.1.1
nameserver 2404:c805:e00:4c00:c256:27ff:feca:219c
```

```
[tom@localhost ~]$ cat /etc/systemd/network/enp3s0.network 
[Match]
Name=enp3s0
[Network]
DHCP=yes
IPv6AcceptRouterAdvertisements=no

[tom@localhost ~]$ grep nameserver /run/systemd/resolve/resolv.conf 
nameserver 192.168.1.1
```

**Comment by tomty89** [comment]
For the record, we do NOT want to just patch link_acquire_ipv6_conf() like this:

```
diff --git a/src/network/networkd-link.c b/src/network/networkd-link.c
index 9ac0b47..3980e63 100644
--- a/src/network/networkd-link.c
+++ b/src/network/networkd-link.c
@@ -1446,6 +1446,8 @@ static int link_acquire_ipv6_conf(Link *link) {

                 log_link_debug(link, "Acquiring DHCPv6 lease");

+                dhcp6_request_address(link);
+
                 r = sd_dhcp6_client_set_local_address(link->dhcp6_client, &link->ipv6ll_address);
                 if (r < 0 && r != -EBUSY)
                         return log_link_warning_errno(link, r, "Could not set IPv6LL address in DHCP client: %m");
```

If we do so, for router that provides "stateless DHCPv6", the user must have `DHCP=` set to `no` or `ipv4` for it to work:

```
[tom@localhost ~]$ rdisc6 enp3s0 | grep -i stateful
Stateful address conf.    :           No
Stateful other conf.      :          Yes
```

```
[tom@localhost ~]$ cat /etc/systemd/network/enp3s0.network 
[Match]
Name=enp3s0
[Network]
DHCP=ipv4

[tom@localhost ~]$ grep nameserver /run/systemd/resolve/resolv.conf 
nameserver 192.168.1.1
nameserver 2404:c805:e00:4c00:c256:27ff:feca:219c
```

```
[tom@localhost ~]$ cat /etc/systemd/network/enp3s0.network 
[Match]
Name=enp3s0
[Network]
DHCP=yes

[tom@localhost ~]$ grep nameserver /run/systemd/resolve/resolv.conf 
nameserver 192.168.1.1
```

Vice versa, if the router provides "stateful DHCPv6", with the current code, DHCPv6 may not work as expected unless `DHCP=` is set to `no` or `ipv4`.

**Comment by teg** [comment]
Hm... so I do think that what you propose makes a lot of sense. But let's first consider the reason it is the way it is:

Why allow DHCP=ipv6 to explicitly enable DHCP without waiting for router advertisements? It is true that this is unlikely to make much sense, though if you have a setup where you KNOW you will need DHCPv6  and you want to reduce the latency of setup (which is already pretty high in IPv6), then you may want these things to be configured in parallel. I'd be ok with forgoing this usecase though, as the benefit seems rather small. Another case to consider is if there is no gateway on the network, so no router, only the local link with its on-link prefixes. You COULD in principle then only have a DHCPv6 server running to hand out the addresses, and configure the on-link prefixes manually. That would work. Would it be a crazy setup? Yes.

Why do we perform DHCPv6 after we find no routers, if nothing else is configured explicitly? This is sort of crazy to me, the only way this could actually work is if the on-link routes were configured manually (as above), but in that case the DHCPv6 client could also have been enabled manually (again, as above). It does appear to be mandated by the RFC though, so we should have some proper understanding of why that is the way it is before we drop this behavior [0](https://tools.ietf.org/html/rfc2462#section-5.5.2>.).

**Comment by teg** [comment]
@dvdhrm informs me that he uses his modem in the way I describe above: only DHCP and no RA with a statically configured route to his modem. Whatever drugs @dvdhrm and his modem manufacturer are sharing, I want some...

**Comment by tomty89** [comment]
@teg:

> It does appear to be mandated by the RFC though, so we should have some proper understanding of why that is the way it is before we drop this behavior.

Well [RFC 2462](https://tools.ietf.org/html/rfc2462#section-5.5.2) has been obsoleted by [RFC 4862](https://tools.ietf.org/html/rfc4862#section-5.5.2). The new one does not seem to mandate (or even recommend) anything about "starting DHCPv6 when Router Discovery timed out" anymore. Instead it just states the possibility of DHCPv6 without RA (and how this quirky stuff can be done, as you've mentioned).

So I guess we should still allow users to explicitly start the DHCPv6 client, but networkd should just do nothing if it does not receive any RA _and_ DHCPv6 is not explicitly enabled with an intended mode specified.

And we'll need to split `DHCP=`:

```
DHCPv6=stateful|stateless|unset ; DHCPv4=yes|no

DHCPv6=unset:
RA flags decide whether DHCPv6 client should be started in stateful or stateless mode;
if no RA received (timed out), do nothing.

DHCPv6=stateful:
Start DHCPv6 client in in stateful mode before/without router discovery.
(i.e. information_request=false)

DHCPv6=stateless
Start DHCPv6 client in stateless mode before/without router discovery.
(i.e. information_request=true)
```

However, if we are going to allow users to enable DHCPv6 explicitly, we also need to consider, when an RA is received:
- Should we restart the DHCPv6 client according to the RA flags?
- Should we stop the DHCPv6 client if the RA flags state that no stateful configuration is provided at all (["M:0, O:0"](https://tools.ietf.org/html/rfc5175))?

**Comment by teg** [comment]
Ah, that's a good catch. Yeah, then we could get away with not starting DHCP after the timeout. I'd be happy to take a patch for that.

If DHCPv6 is enabled explicitly, then we should not let it be affected by RA at all. If you enable DHCPv6 explicitly, it must be because you are doing something special in an environment requiring manual configuration anyway, so we should not do any magic there, but expect the admin to have set up everything the way it needs to be.

Changing the config option, I would keep as a separate PR. The change you are proposing is to allow selecting manually between stateful or stateless info. Could it be that this is overkill? Would be nice if we could keep it as simple as it is now I guess. I'm not entirely convinced either way, so let's make that a separate PR and a separate debate? @poettering you originally suggested we merge DHCPv4 and DHCPv6 into one, thoughts on this proposal?

**Comment by tomty89** [comment]
But do we have any ground to say that we could only want stateful (or stateless) DHCPv6 when we have it enable explicitly? Cause apparently requesting with the wrong mode would prevent it from working (or perhaps I just have a router that sucks?) In that case if we also choose not to let RA affect it at all once it is enabled explicitly, users may need to be very careful NOT to have it enabled. With the current way we do it we would need to bare in mind to stick with `DHCP=ipv4` instead of `=yes`.

It might indeed be a bit overkill, that's why I proposed originally we should solely rely on RA to trigger DHCPv6. It seems to me that's the only rational way to avoid being overkill.

**Comment by poettering** [comment]
Hmm, so what about this:

```
[Network]
DHCP=yes|no|ipv4|ipv6          -- this we already have
[DHCP]
Mode=auto|ra|yes|stateless     -- this would be new, and default to "auto"
```

Or something like that. `Mode=` would be new, and its setting `auto` would be "use if RA says so, or if RA doesn't respond".  `ra` would be "use if RA says so, but only then". `yes` would be "do dhcp6 immediately in stateful mode". And finally `stateless` would be "do dhcp6 immediately in stateless mode".

Does this make sense and cover all cases?

**Comment by tomty89** [comment]
We don't need `auto`, since we agreed that networkd should do nothing about DHCPv6 after  router discovery timeout (https://github.com/systemd/systemd/commit/acac5b2f512e03830df06df3fc30aff28d01d6b3).

And no, it still doesn't really make sense. What we need is merely to feed networkd with a parameter (`stateful` or `stateless`), and when it is fed with that, we can safely assume that users want DHCPv6 to be started explicitly and RA flags to be ignored. So we don't need the "v6 part" of `DHCP=` at all anyway. (Unless we are going to allow users to disable DHCPv6 explicitly, including the RA triggering I mean)

Sure we can make `Mode=` accept only `stateful` or `stateless` and make one of them as the default when not set, but why do we want a default mode anyway?

**Comment by poettering** [comment]
@tomty89 hmm, what's so wrong about trying DHCPv6 after RA timeout? I mean, as discussed @dvdhrm appears to have a router like that, and it would be good to cover that by default.

And again, I am pretty sure there should be a way to disable DHCP6 without turning off RA too. I mean, RA covers a lot, and is fully sufficient if all you need is DNS and an IP, hence, if RA includes that, and sets the "O" bit (i.e. meaning that DHCPv6 might have more data), then there might be a good reason not to bother with that...

**Comment by tomty89** [comment]
Because we would need to tell the client what mode it should start in if RA timeout, and when we do, we don't need to wait until RA timeout.

I actually like the idea that we provide an absolute no option for DHCPv6. I just thought you guys don't like it somehow (because of the way it has been so far).

**Comment by poettering** [comment]
@tomty89 well, my guess is that most OSes set DHCPv6 to stateful mode, and we should probably prefer that too, no? Hence Mode=auto would make a great default: use stateless DHCP if indicated by RA or if RA doesn't respond. That setting should work for most people then?

**Comment by tomty89** [comment]
I don't really quite follow here. You mean like (with `DHCP=ipv6|yes`):

`Mode=auto`: listen to RA if received, do stateful DHCPv6 if timeout
`Mode=ra`: listen to RA if received, do nothing if timeout
`Mode=yes`: do stateful DHCPv6 right away and ignore RA
`Mode=stateless`: do stateless DHCPv6 right away and ignore RA

If that's what you mean, well, personally I am not a big fan of having a "biased" default, especially when there's not really a strong reasoning to do so; but it's acceptable.

Still I don't see much point of starting DHCPv6 if timeout. IIUC, according to what @teg(@dvdhrm) and the RFC described, DHCPv6 without RA won't just "work out-of-the-box". You'll need a some quirky manual configuration anyway. In that case, if someone need to start DHCPv6 without RA, will he ever want networkd to try RA first?

If there's really a need for this "try and fallback" mechanism, don't we need to make the option balanced and complete:

`Mode=auto_stateless`: listen to RA if received, do stateless DHCPv6 if timeout

**Comment by poettering** [comment]
Well, I think it makes sense to have a "biased" default, absolutely. I mean, most operating systems for the desktop and for embedded will probably default to stateful dhcp6 or so, and I think given that this is the way it is, we really should be biased towards that...

**Comment by tomty89** [comment]
> most operating systems for the desktop and for embedded will probably default to stateful dhcp6 or so, and I think given that this is the way it is, we really should be biased towards that...

Well, if you say so.

But at the very least, I don't think it should treated as a single question like this:
- Do we want to start stateful DHCPv6 when RA timeout?

But separate questions like these:
- Do we want to start DHCPv6 when RA timeout?
- If so, do we want to do that by default?
- If so, which mode do we default to?

So I while I consider a "biased" default "okay", having `Mode=auto` but not `Mode=auto_stateless` is not an option to me (that said, it's still your call).

If five options is considered overwhelming or overkill or so, I would rather we do not allow users to explicitly start DHCPv6 right away and ignore RA from the beginning:

`Mode=auto` (default): listen to RA if received, do stateful DHCPv6 if timeout
`Mode=auto_stateless`: listen to RA if received, do stateless DHCPv6 if timeout
`Mode=ra_only`: listen to RA if received, do nothing if timeout

**Comment by tomty89** [comment]
Btw, what about making them three booleans:

```
[DHCP]
# when true, start DHCPv6 right away and ignore RA flags
EarlyStart=yes|no (default: no)
# when true, start DHCPv6 when no RA is received (if EarlyStart=no)
RATimeout=yes|no (default: yes)
# when true, start "stateless" DHCPv6; used only by EarlyStart=yes or RATimeout=yes
InfoReq=yes|no (default: no, a.k.a "stateful" DHCPv6)
```

I think that makes coding easier and may _possibly_ be less confusing to users.

**Comment by poettering** [comment]
I don't like settings that are redundant. Specifically: EarlyStart=yes with RATimeout=yes at the same time makes no sense, does it?

**Comment by tomty89** [comment]
Hmm, maybe we just leave out `RATimeout=`? With the possibilities to start DHCPv6 early/explicitly and to disable RA triggering completely without dropping RA, I guess the option is a bit redundant anyway, no matter in what way we put it.

**Comment by yrro** [comment]
Perhaps an issue I have is covered by this: DHCPv6 is triggered even though the RA packets I am receiving have neither the Managed Address Configuration or Other Stateful Confiruation bits set. In this case, DHCPv6 queries should not be sent! I can file a separate issue if there is no overlap with this one, but I'm not sure.
